lessons/r/ggplot2/ggplot2.Rmd

```diff
@@ -0,0 +1,423 @@
+
+```{r}
+ggplot(diamonds, aes(x = carat)) +
+  geom_histogram(binwidth = 0.05)
+```
+
+### Bar plots with `dplyr` and `ggplot` - using grouped `summarise` operations
+
+Earlier, we saw that the `clarity` variable seemed to be associated with the carat of a given diamond.
+Let's investigate this further by calculating the mean carat for each value of `clarity`.
+This can be done using a grouped `summarise` operation via `dplyr`:
+
+```{r}
+diamonds %>% 
+  group_by(clarity) %>% 
+  summarise(mean_carat = mean(carat),
+            sd_carat = sd(carat))
+```
+
+Let's pipe this into a ggplot call with `geom_bar`. We have to be careful to specify `stat = 'identity'` as an
+argument to `geom_bar`. This is because `geom_bar` will otherwise try to apply some sort of transformation instead
+of trying to plot single values, like the ones we have above.
+
+```{r}
+diamonds %>% 
+  group_by(clarity) %>% 
+  summarise(mean_carat = mean(carat),
+            sd_carat = sd(carat)) %>% 
+  ggplot(aes(x = clarity, y = mean_carat)) +
+  geom_bar(stat = 'identity')
+```
+
+We also calculated the standard deviation above using `sd()` within `summarise`. This can
+be used to generate an errorbar with `geom_errorbar()`:
+
+```{r}
+diamonds %>% 
+  group_by(clarity) %>% 
+  summarise(
+    mean_carat = mean(carat),
+    sd_carat = sd(carat)) %>% 
+  ggplot(aes(x = clarity, y = mean_carat)) +
+  geom_bar(stat = 'identity') +
+  geom_errorbar(aes(ymin = mean_carat - sd_carat, 
+                    ymax = mean_carat + sd_carat))
+```
+
+### Boxplots
+
+Although the above plot is in line with the pattern we observed in the scatter plot at the start of the lesson,
+we may want to go a step further and visualize the distribution of `carat` for each value of `clarity` with a boxplot.
+These are rather straightforward to make:
+
+```{r}
+ggplot(diamonds, aes(x = clarity, y = carat)) +
+  geom_boxplot()
+```
+
+## Model fits with `geom_smooth`
+
+Fitting models to data is a common means of investigating whether certain relationships are at play in a given dataset.
+`ggplot2` makes visualizing these easy by largely consolidating line-fitting into a single geom - `geom_smooth`.
+
+By default, `geom_smooth` will automatically select what it thinks is the best model for your dataset. However, rarely
+will it select a linear fit. To draw a linear fit, we have to specify it:
+
+```{r}
+ggplot(diamonds, aes(x = carat, y = price)) +
+  geom_point(alpha = 0.5) +
+  geom_smooth(method = 'lm')
+```
+
+Adding the color aesthetic we had previously will cause lines to be fit to each individual group:
+
+```{r}
+ggplot(diamonds, aes(x = carat, y = price, color = clarity)) +
+  geom_point(alpha = 0.5) +
+  geom_smooth(method = 'lm')
+```
+
+To still fit a line across the entire dataset, specify the color aesthetic _within_ `geom_point`, and not globally.
+
+```{r}
+ggplot(diamonds, aes(x = carat, y = price)) +
+  geom_point(aes(color = clarity), alpha = 0.5) +
+  geom_smooth(method = 'lm')
+
```

Comment:
Maybe worth mentioning (but not showing due to time constraints) that you can also specify non-linear fits (e.g. quadratic, cubic) and LOESS fits.